storage/innobase/read/read0read.cc

```diff
@@ -262,4 +307,552 @@ void trx_sys_t::clone_oldest_view(ReadViewBase *view) const
+    return;
+  }
+  if (is_scn(id) || id == current_id)
+  {
+    /* Already be filled with scn, do nothing */
+    return;
+  }
+
+  trx_id_t scn= get_scn(id, index, row_get_rec_roll_ptr(rec, index, offsets));
+
+  if (scn == 0 || scn == TRX_ID_MAX)
+  {
+    return;
+  }
+
+  set_scn(mtr, block, rec, offset, id, scn);
+}
+
+void SCN_Mgr::add_lazy_cursor(buf_block_t *block, rec_t *rec,
+                              ulint trx_id_offset, trx_id_t id, trx_id_t scn,
+                              table_id_t table_id)
+{
+  ut_ad(srv_use_scn);
+  if (fsp_is_system_temporary(block->get_space_id()))
+  {
+    return;
+  }
+
+  if (block->add_lazy_cursor(rec, trx_id_offset, id, scn) == false)
+  {
+    return;
+  }
+  ut_a(rec > block->page.frame);
+
+  /* Add page number to the set */
+  page_id_t page_id{block->get_space_id(), block->get_page_no()};
+  uint64_t compact_id= page_id.raw();
+  uint64_t slot= compact_id % m_max_cleanout_threads;
+
+  m_cleanout_workers[slot]->add_page(compact_id, table_id);
+}
+
+void SCN_Mgr::view_task()
+{
+  ut_a(srv_use_scn);
+  if (m_view_task_mutex.try_lock() == false)
+  {
+    return;
+  }
+  trx_id_t id= 0;
+  trx_id_t no= 0;
+  trx_sys.get_min_trx_id_no(id, no);
+  m_safe_limit_no= no;
+  m_min_active_id= id;
+  m_view_task_mutex.unlock();
+}
+
+void SCN_Mgr::batch_write(buf_block_t *block, mtr_t *mtr)
+{
+  ut_ad(srv_use_scn);
+  LazyCleanoutRecs lrecs;
+  lrecs.clear();
+  block->copy_and_free(lrecs);
+
+  for (auto &lrec : lrecs)
+  {
+    rec_t *rec= lrec.first;
+    ulint trx_id_offset= std::get<0>(lrec.second);
+    trx_id_t trx_id= std::get<1>(lrec.second);
+    trx_id_t scn= std::get<2>(lrec.second);
+    set_scn(mtr, block, rec, trx_id_offset, trx_id, scn);
+  }
+}
+
+void SCN_Mgr::cleanout_task(uint32_t slot)
+{
+  ut_a(srv_use_scn);
+  CleanoutWorker *worker= m_cleanout_workers[slot];
+  MDL_ticket *mdl_ticket;
+  THD *thd= worker->get_thd();
+  while (!m_abort.load())
+  {
+    PageSets pages;
+    pages.clear();
+    worker->take_pages(pages);
+    if (pages.empty())
+    {
+      break;
+    }
+    /* Process pages that need to be modified */
+    for (auto &val : pages)
+    {
+      if (m_abort.load())
+      {
+        break;
+      }
+      page_id_t page_id(val.first);
+      table_id_t table_id= val.second;
+      mdl_ticket= nullptr;
+      /* Prevents the tablespace from being dropped during the operation */
+      dict_table_t *table= dict_table_open_on_id(
+          table_id, false, DICT_TABLE_OP_NORMAL, thd, &mdl_ticket);
+      if (table == nullptr)
+      {
+        continue;
+      }
+      fil_space_t *space= fil_space_get(page_id.space());
+      ulint zip_size= 0;
+      ut_a(space);
+      zip_size= space->zip_size();
+      mtr_t mtr;
+      buf_block_t *block= nullptr;
+      mtr_start(&mtr);
+      mtr.set_in_scn_cleanout();
+      block= buf_page_get_gen(page_id, zip_size, RW_X_LATCH, nullptr,
+                              BUF_GET_IF_IN_POOL, &mtr, nullptr);
```

Comment:
`BUF_GET_IF_IN_POOL` does not require `zip_size`.

Reply:
OK.